File: apps/frontend/src/features/plans/schemas.ts
import { z } from "zod";

export const PlanRoleProgressSchema = z.object({
  role: z.string().min(1),
  totalCheckpoints: z.number().int().min(0),
  doneCheckpoints: z.number().int().min(0),
});

export const PlanOverallProgressSchema = z.object({
  totalCheckpoints: z.number().int().min(0),
  doneCheckpoints: z.number().int().min(0),
  percentComplete: z.number().int().min(0).max(100),
});

export const PlanCheckpointSchema = z.object({
  timestamp: z.string().min(1),
  role: z.string().min(1),
  checkpointId: z.string().min(1),
  state: z.string().min(1),
  message: z.string(),
});

export const OpenSpecPlanSummarySchema = z.object({
  slug: z.string().min(1),
  title: z.string().min(1),
  status: z.string().min(1),
  updatedAt: z.string().datetime({ offset: true }),
  roles: z.array(PlanRoleProgressSchema).default([]),
  overallProgress: PlanOverallProgressSchema,
  currentCheckpoint: PlanCheckpointSchema.nullable(),
});

export const OpenSpecPlansResponseSchema = z.object({
  entries: z.array(OpenSpecPlanSummarySchema).default([]),
});

export const OpenSpecPlanRoleDetailSchema = z.object({
  role: z.string().min(1),
  totalCheckpoints: z.number().int().min(0),
  doneCheckpoints: z.number().int().min(0),
  tasksMarkdown: z.string(),
  checkpointsMarkdown: z.string().nullable(),
});

export const OpenSpecPlanDetailSchema = z.object({
  slug: z.string().min(1),
  title: z.string().min(1),
  status: z.string().min(1),
  updatedAt: z.string().datetime({ offset: true }),
  summaryMarkdown: z.string(),
  checkpointsMarkdown: z.string(),
  roles: z.array(OpenSpecPlanRoleDetailSchema).default([]),
  overallProgress: PlanOverallProgressSchema,
  currentCheckpoint: PlanCheckpointSchema.nullable(),
});

export const PlanRuntimeAgentSchema = z.object({
  name: z.string().min(1),
  role: z.string().nullable(),
  model: z.string().nullable(),
  status: z.string().nullable(),
  startedAt: z.string().nullable(),
  updatedAt: z.string().nullable(),
  source: z.string().min(1),
  authoritative: z.boolean(),
});

export const PlanRuntimeEventSchema = z.object({
  ts: z.string().min(1),
  kind: z.string().min(1),
  message: z.string(),
  agentName: z.string().nullable(),
  role: z.string().nullable(),
  model: z.string().nullable(),
  status: z.string().nullable(),
  source: z.string().min(1),
  authoritative: z.boolean(),
});

export const PlanRuntimeErrorSchema = z.object({
  timestamp: z.string().min(1),
  code: z.string().nullable(),
  message: z.string(),
  source: z.string().nullable(),
  recoverable: z.boolean().nullable(),
});

export const OpenSpecPlanRuntimeSchema = z.object({
  available: z.boolean(),
  sessionId: z.string().nullable(),
  correlationConfidence: z.string().nullable(),
  mode: z.string().nullable(),
  phase: z.string().nullable(),
  active: z.boolean(),
  updatedAt: z.string().datetime({ offset: true }).nullable(),
  agents: z.array(PlanRuntimeAgentSchema).default([]),
  events: z.array(PlanRuntimeEventSchema).default([]),
  lastCheckpoint: PlanCheckpointSchema.nullable(),
  lastError: PlanRuntimeErrorSchema.nullable(),
  canResume: z.boolean(),
  partial: z.boolean(),
  staleAfterSeconds: z.number().int().positive().nullable(),
  reasons: z.array(z.string()).default([]),
  unavailableReason: z.string().nullable(),
});

export type OpenSpecPlanSummary = z.infer<typeof OpenSpecPlanSummarySchema>;
export type OpenSpecPlansResponse = z.infer<typeof OpenSpecPlansResponseSchema>;
export type OpenSpecPlanDetail = z.infer<typeof OpenSpecPlanDetailSchema>;
export type OpenSpecPlanRuntime = z.infer<typeof OpenSpecPlanRuntimeSchema>;

File: app/modules/plans/schemas.py
from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.modules.shared.schemas import DashboardModel


class PlanRoleProgress(DashboardModel):
    role: str
    total_checkpoints: int
    done_checkpoints: int


class PlanOverallProgress(DashboardModel):
    total_checkpoints: int
    done_checkpoints: int
    percent_complete: int


class PlanCheckpoint(DashboardModel):
    timestamp: str
    role: str
    checkpoint_id: str
    state: str
    message: str


class OpenSpecPlanSummary(DashboardModel):
    slug: str
    title: str
    status: str
    created_at: datetime
    updated_at: datetime
    summary_markdown: str
    roles: list[PlanRoleProgress]
    overall_progress: PlanOverallProgress
    current_checkpoint: PlanCheckpoint | None


class OpenSpecPlansResponse(DashboardModel):
    entries: list[OpenSpecPlanSummary]


class OpenSpecPlanRoleDetail(DashboardModel):
    role: str
    total_checkpoints: int
    done_checkpoints: int
    tasks_markdown: str
    checkpoints_markdown: str | None


class PlanPromptItem(DashboardModel):
    id: str
    title: str
    content: str
    source_path: str


class PlanPromptBundle(DashboardModel):
    id: str
    title: str
    source_path: str
    prompts: list[PlanPromptItem] = Field(default_factory=list)


class OpenSpecPlanDetail(DashboardModel):
    slug: str
    title: str
    status: str
    created_at: datetime
    updated_at: datetime
    summary_markdown: str
    checkpoints_markdown: str
    roles: list[OpenSpecPlanRoleDetail]
    overall_progress: PlanOverallProgress
    current_checkpoint: PlanCheckpoint | None
    prompt_bundles: list[PlanPromptBundle] = Field(default_factory=list)


class PlanRuntimeAgent(DashboardModel):
    name: str
    role: str | None
    model: str | None
    status: str | None
    started_at: str | None
    updated_at: str | None
    source: str
    authoritative: bool


class PlanRuntimeEvent(DashboardModel):
    ts: str
    kind: str
    message: str
    agent_name: str | None
    role: str | None
    model: str | None
    status: str | None
    source: str
    authoritative: bool


class PlanRuntimeError(DashboardModel):
    timestamp: str
    code: str | None
    message: str
    source: str | None
    recoverable: bool | None


class OpenSpecPlanRuntime(DashboardModel):
    available: bool
    session_id: str | None
    correlation_confidence: str | None
    mode: str | None
    phase: str | None
    active: bool
    updated_at: datetime | None
    agents: list[PlanRuntimeAgent]
    events: list[PlanRuntimeEvent]
    last_checkpoint: PlanCheckpoint | None
    last_error: PlanRuntimeError | None
    can_resume: bool
    partial: bool
    stale_after_seconds: int | None
    reasons: list[str]
    unavailable_reason: str | None
